### ISSUE TITLE
JQ3.0 RC1, Deferred.notify / .progress callback getting the wrong params

### ISSUE THREAD

**Issue by VanPuyveldeKevin** [jquery/jquery#3129]
When calling Deferred.notify([params]) it seems like the progress callback on the given promise ends up in an unexpected (incorrect?) situation.
It's receiving undefined as the first param and the params array object in the second.
Basically ... the 'notifyWith' sig (context, [params]).

I would expect the '.notify([params])' to still behave the same as in 1.x/2.x and give the inline progress callback with the given params.

```js
return _retrieve()
            .then(function (data, textStatus, jqXhr)
            {
                deferred.notify(data, textStatus, jqXhr);
                if (data["@odata.nextLink"])
                {
                }
                else
                {
                    deferred.resolve(data, textStatus, jqXhr);
                    return data;
                }
            })

.then(function (data, textStatus, jqXhr)
    {
        assert.ok(true, "Records created");
        return $a.WebAPI.retrieve();
    })
.progress(function (data, textStatus, jqXhr)
    {
        assert.ok(true, "Retrieved a page");        
    })
```

The progress callback here gets undefined in 'data' (so no context) and the 'params' object of the notifyWith in the 'textStatus' (with the 3 given params as [0], [1] and [2]).

**EDIT** (by @mgol): fixed the Markdown for proper code formatting.

**Comment by mgol** [comment]
This seems to be a duplicate of #3062 so let's continue there (unless it's not, then please comment & I'll reopen).

**Comment by mgol** [comment]
And thanks for the report. :)

**Comment by mgol** [comment]
@VanPuyveldeKevin Would you be interested in preparing a pull request? The fix suggested in #3062 is simple, most of the work would be in adding a proper test case, we could start from your code from this report.

We can guide you through the process & help if you get stuck. :)

**Comment by mgol** [comment]
@VanPuyveldeKevin OK, I'll be preparing a PR since I want to get it done for 3.0.0.

I see that I'm missing context in your example. What exactly is the `deferred` you're using in the code snippet? How does it relate to what `_retrieve()` returns?

**Comment by mgol** [comment]
PR: #3150

**Comment by VanPuyveldeKevin** [comment]
Basically I'm using a new $.Deferred() to implement the progress notification.
This because the $.ajax returns me a promise which does not implement that function.

Of course the above example was very simplified from the original implementation.
In the implementation I'm using the '.then' to implement deferred recursion (odata, check next page, do new retrieve). Of course it's a nifty feature to already be able to get some intermediate results while loading, hence progress.

In below example ... callers of retrieve all pages can get progress containing each page. Basically retrieved using $.ajax (hence the ajax success sig).

Something like this:

    var retrieveAllPages = function ()
    {
        var deferred = $ajq.Deferred();

        function internalRetrieve(webApiRequestOptions)
        {
            return _retrieveMultiple(webApiRequestOptions)
            .then(function (data, textStatus, jqXhr)
            {
                deferred.notify(data, textStatus, jqXhr);
                if (data["@odata.nextLink"])
                {
                    return internalDoRequest(data["@odata.nextLink"], webApiRequestOptions, data.value);
                }
                else
                {
                    deferred.resolve(data, textStatus, jqXhr);
                    return data;
                }
            })
            .fail(function (jqXhr, textStatus, errorThrown)
            {
                deferred.reject(jqXhr, textStatus, errorThrown);
            });
        };

        function internalDoRequest(url, webApiRequestOptions, items)
        {
            return doRequest({ url: url }, webApiRequestOptions)
            .then(function (data, textStatus, jqXhr)
            {
                deferred.notify(data, textStatus, jqXhr);
                if (data["@odata.nextLink"])
                {
                    return internalDoRequest(data["@odata.nextLink"], webApiRequestOptions, items.concat(data.value));
                }
                else
                {
                    // We reached the last page, so we'll return the full object now (of course WITH all the values)
                    data.value = items.concat(data.value);
                    deferred.resolve(data, textStatus, jqXhr);
                    return data;
                }
            })
            .fail(function (jqXhr, textStatus, errorThrown)
            {
                deferred.reject(jqXhr, textStatus, errorThrown);
            });
        };

        var requestOptions = helpers.inlineSigToOptions();

        // Start the retrieval. Retrieving the first page before starting an async resursive retrieval process.
        internalRetrieveMultiple(requestOptions);

        return deferred.promise();
    };

**Comment by mgol** [comment]
@VanPuyveldeKevin OK, thanks. Could you check if https://dl.dropboxusercontent.com/u/10200888/jquery/jquery-gh-3150.js resolves your issue?

**Comment by mgol** [comment]
@VanPuyveldeKevin Would you be able to confirm? I'd like to include this fix in 3.0.0 but I'd like some confirmation that this is exactly what broke for you. :)

**Comment by VanPuyveldeKevin** [comment]
Alright ... ran it through the QUnit test and it's closer, but still seems to be a breaking change compared to the earlier major version 1.14 (latest).

As you can see in the example, it's a recursive '.then'. Example with a page size of 2 and total of 3 for ease. So 2 pages.

- InternalRetrieve does a data retrieve, gets 2 items in the result, does progress with those 2 items -> **OK**, '.progress' callback gets those 2 results.
- Recursive InternalDoRequest does the retrieve for page 2, gets 1 item in the result, does progress with that 1 item -> **NOK**, '.progress' callback does not get 1 result, but all 3 results somehow.

**DETAILS:**
- I seem to end up (in debug flow) in the 'internalDoRequest, else flow' first. Basically, concat of the last page and full deferred resolving are done before ending up in the progress callback.
  -> As a result the last page progress callback in this situation gets the full result instead of a single page retrieved

**NOTE:** 
In previous versions the exact same approach does in fact end up in the '.progress' callback with that last page result (so only the 1 item).

**THOUGHTS:**
Personally I think the current operation mode is not ideal since the progress callback is often used to start binding data (for example using frameworks like knockoutjs) while still loading data.
Normally this is of course not with a simple 3 records.
So when the last page is loaded all data that's already bound, to let's say a grid, would be re-added all at once, adding duplicates for pretty much the whole list _except_ for the final page.

**Comment by mgol** [comment]
@VanPuyveldeKevin The code you posted has some extra details that are not relevant to this issue; could you try to reduce the test case and post it in a working state at https://jsfiddle.net/ with clear description of what you expect to happen? Something like the #3062's example in https://jsfiddle.net/zwavzvpx/.

**Comment by VanPuyveldeKevin** [comment]
Well now ... tried reproducing it in a simple case, but did not succeed for now. In the above (realistic) case the last '.progress' callback isn't triggered. I assume that means the callback queueing contains the breaking change then. Since in that case I reach the deferred resolve first, then the progress callback is triggered (meaning the order is off).
However, in a simple case (without $.ajax) it doesn't seem to occurr even if I more or less replicate the logic.

Below I setup 2 example test cases, both of which seem to not have the issue (.notify immediately brings the flow into the .progress callback).
As I see it the biggest difference is the $.ajax deferred not being in the mix here (since I fake actual data retrieval here).

_(Formatting doesn't seem to work too well though)_
`
var retrievePage = function ()
{
    var deferred = $.Deferred();
    var pageItems = ["Item1", "Item2"];

    setTimeout(function () { deferred.resolve(pageItems); }, 500);

    return deferred.promise();
};

var retrieveAllPages = function ()
{
    var deferred = $.Deferred();

    function internalRetrieveAllPages(items, pageNumber)
    {
        return retrievePage()
        .then(function (pageItems)
        {
            deferred.notify(pageItems, pageNumber);
            if (pageNumber == 1)
            {
                return internalRetrieveAllPages(items.concat(pageItems), pageNumber + 1);
            }
            else
            {
                var finalResult = items.concat(pageItems);
                deferred.resolve(finalResult);
                return finalResult;
            }
        })
        .fail(function (jqXhr, textStatus, errorThrown)
        {
            deferred.reject(jqXhr, textStatus, errorThrown);
        });
    };

    internalRetrieveAllPages([], 1);

    return deferred.promise();
};

var retrieveAllPages2 = function ()
{
    var deferred = $.Deferred();

    function internalRetrieveFirstPage()
    {
        var pageNumber = 1;
        return retrievePage()
        .then(function (pageItems)
        {
            deferred.notify(pageItems, pageNumber);
            if (pageNumber == 1)
            {
                return internalRetrieveNextPagesRecursive(pageItems, pageNumber + 1);
            }
            else
            {
                var finalResult = items.concat(pageItems);
                deferred.resolve(finalResult);
                return finalResult;
            }
        })
        .fail(function (jqXhr, textStatus, errorThrown)
        {
            deferred.reject(jqXhr, textStatus, errorThrown);
        });
    };

    function internalRetrieveNextPagesRecursive(items, pageNumber)
    {
        return retrievePage()
        .then(function (pageItems)
        {
            deferred.notify(pageItems, pageNumber);
            if (pageNumber <= 3)
            {
                return internalRetrieveNextPagesRecursive(items.concat(pageItems), pageNumber + 1);
            }
            else
            {
                var finalResult = items.concat(pageItems);
                deferred.resolve(finalResult);
                return finalResult;
            }
        })
        .fail(function (jqXhr, textStatus, errorThrown)
        {
            deferred.reject(jqXhr, textStatus, errorThrown);
        });
    };

    internalRetrieveFirstPage([], 1);

    return deferred.promise();
};

QUnit.test("RetrieveAllPages", function (assert)
{
    var done = assert.async();
    retrieveAllPages()
    .progress(function (page, pageNumber)
    {
        assert.ok(page.length == 2, "Progress - Retrieved a page. Found the right amount of accounts relative the page size. Found: " + page.length);
    })
    .done(function (allPages)
    {
        assert.ok(allPages.length == 4, "Done - Retrieved all pages. Found the right amount of accounts relative the page size. Found: " + allPages.length);
    })
    .always(function ()
    {
        done();
    });
});

QUnit.test("RetrieveAllPages2", function (assert)
{
    var done = assert.async();
    retrieveAllPages2()
    .progress(function (page, pageNumber)
    {
        assert.ok(page.length == 2, "Progress - Retrieved a page. Found the right amount of accounts relative the page size. Found: " + page.length);
    })
    .done(function (allPages)
    {
        assert.ok(allPages.length == 8, "Done - Retrieved all pages. Found the right amount of accounts relative the page size. Found: " + allPages.length);
    })
    .always(function ()
    {
        done();
    });
});`
